pkg/globalinit/catalog_cache_init.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package globalinit
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os/exec"
+	"sort"
+
+	"github.com/pkg/errors"
+
+	"github.com/vmware-tanzu/tanzu-cli/pkg/catalog"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/cli"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/lastversion"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/pluginsupplier"
+)
+
+// This global initializer checks if the last executed CLI version is < 1.3.0.
+// If so it refreshes the plugin catalog to make sure any remapping data is in the catalog.
+
+func init() {
+	RegisterInitializer("Plugin Info Catalog Initializer", triggerForPreCommandRemapping, refreshPluginCatalog)
+}
+
+func triggerForPreCommandRemapping() bool {
+	// If the last executed CLI version is < 1.3.0, we need to refresh the plugin catalog.
+	return lastversion.GetLastExecutedCLIVersion() == lastversion.OlderThan1_3_0
+}
+
+// refreshPluginCatalog reads the info from each installed plugin
+// and updates the plugin catalog with the latest info.
+// We need to do this to make sure the command re-mapping data is in the cache.
+func refreshPluginCatalog(outStream io.Writer) error {
+	plugins, err := pluginsupplier.GetInstalledPlugins()
+	if err != nil {
+		return err
+	}
+	sort.Sort(cli.PluginInfoSorter(plugins))
+
+	c, err := catalog.NewContextCatalogUpdater("")
+	if err != nil {
+		return err
+	}
+	defer c.Unlock()
+
+	for i := range plugins {
+		fmt.Fprintf(outStream, "Refreshing plugin %s...\n", plugins[i].Name)
+		if pInfo, err := refreshPluginInfo(&plugins[i], plugins[i].InstallationPath); err == nil {
+			_ = c.Upsert(pInfo)
```

Comment:
To avoid any possible race condition, we might have to create a new instance of catalog with `NewContextCatalogUpdater` every time and release the lock before processing the next plugin as some plugin might be trying to read catalog cache which might cause deadlock.

Reply:
Thanks, I didn't think of that at all!